File: frontend/de/unika/ipd/grgen/ir/RetypedNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 3.0
 * Copyright (C) 2003-2011 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @version $Id$
 */
package de.unika.ipd.grgen.ir;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import de.unika.ipd.grgen.util.Annotations;
import de.unika.ipd.grgen.util.Retyped;

public class RetypedNode extends Node implements Retyped {

	/**  The original entity if this is a retyped entity */
	protected Node oldNode = null;
	
	/** A list of nodes to be additionally merged into the retyped node  */
	private final List<Node> mergees = new LinkedList<Node>();

	public RetypedNode(Ident ident, NodeType type, Annotations annots,
			boolean maybeDeleted, boolean maybeRetyped, boolean isDefToBeYieldedTo, int context) {
		super(ident, type, annots, null, maybeDeleted, maybeRetyped, isDefToBeYieldedTo, context);
	}

	public Entity getOldEntity() {
		return oldNode;
	}

	public void setOldEntity(Entity old) {
		this.oldNode = (Node)old;
	}

	/** returns the original node in the graph. */
	public Node getOldNode() {
		return oldNode;
	}

	/** Set the old node being retyped to this one */
	public void setOldNode(Node old) {
		this.oldNode = old;
	}

	public boolean isRetyped() {
		return true;
	}

	public void addMergee(Node mergee) {
		mergees.add(mergee);
	}
	
	public List<Node> getMergees() {
		return Collections.unmodifiableList(mergees);
	}
	
	public int getCombinedDependencyLevel() {
		int depLevel = oldNode.getDependencyLevel();
		for(Node mergee : mergees) {
			depLevel = Math.max(depLevel, mergee.getDependencyLevel());
		}
		return depLevel;
	}
}

File: frontend/de/unika/ipd/grgen/ir/pattern/RetypedNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

namespace de.unika.ipd.grgen.ir.pattern
{

using System;
using System.Collections.Generic;

using Entity = de.unika.ipd.grgen.ir.Entity;
using Ident = de.unika.ipd.grgen.ir.Ident;
using NodeType = de.unika.ipd.grgen.ir.model.type.NodeType;
using Annotations = de.unika.ipd.grgen.util.Annotations;
using Retyped = de.unika.ipd.grgen.util.Retyped;

public class RetypedNode : Node, Retyped
{
	/// <summary>
	///  The original entity if this is a retyped entity </summary>
	protected internal Node oldNode = null;

	/// <summary>
	/// A list of nodes to be additionally merged into the retyped node </summary>
	private readonly List<Node> mergees = new List<Node>();

	public RetypedNode(Ident ident, NodeType type, Annotations annots,
			bool maybeDeleted, bool maybeRetyped, bool isDefToBeYieldedTo, int context)
		: base(ident, type, annots, null, maybeDeleted, maybeRetyped, isDefToBeYieldedTo, context)
	{
	}

	public virtual Entity OldEntity
	{
		get
		{
		return oldNode;
		}
		set
		{
		this.oldNode = (Node)value;
		}
	}


	/// <summary>
	/// returns the original node in the graph. </summary>
	public virtual Node OldNode
	{
		get
		{
		return oldNode;
		}
		set
		{
		this.oldNode = value;
		}
	}


	public override bool IsRetyped()
	{
		return true;
	}

	public virtual void AddMergee(Node mergee)
	{
		mergees.Add(mergee);
	}

	public virtual IList<Node> Mergees
	{
		get
		{
		return mergees.AsReadOnly();
		}
	}

	public virtual int CombinedDependencyLevel
	{
		get
		{
		int depLevel = oldNode.DependencyLevel;
		foreach(Node mergee in mergees)
			depLevel = Math.Max(depLevel, mergee.DependencyLevel);
		return depLevel;
		}
	}
}

}
